[inductor] scale down RBLOCK for occupancy

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109275

For large reduction (with large xnumel and rnumel), we potentially need run large number of thread blocks. Occupancy matters here since with larger occupancy we can run more blocks on each SM and we may need less number of waves to run the entire kernel on the GPU.  Number of registers used by each thread can limit the occupancy. For A100, it's safe to say that register usage does not limit occupancy only if each thread use <= 32 registers. This PR leverage this observation and reduce RBLOCK (thus reduce registers used by each thread) if thread usage limit occupancy for large reduction.

The scenario mentioned can happen for the softmax kernel used in transformers. Here are some results get from devgpu:
- PLBartForCausalLM we improve from 1.88x (58.7ms) to 2.00x (55.82ms)
- TrOCRForCausalLM we improve from 1.45x (92.9ms) to 1.51x (89.12ms)


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

shunting314 committed:
[inductor] scale down RBLOCK for occupancy

[ghstack-poisoned]

Chillee commented:
Very nice :)

lezcano commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -165,10 +166,56 @@ def precompile(self, warm_cache_only_with_cc=None):
+                    if total_block <= max_blocks_per_sm * 108:
+                        # <100% occupancy is fine
+                        continue
+                    # make sure rblock is not too small
+                    if rblock <= 64:
+                        continue
```
In this case, wouldn't we want to decrease `XBLOCK`?

shunting314 replied:
yea, we could in order to reduce register usages for each thread. But in this case, we also gonna have more threads and blocks. Occupancy is improved but we may still need the same number of waves to run all the thread blocks.

Also, in most cases I checked, our XBLOCK is usually 1 :) We can not decrease further. I think it's due to this heuristic: https://github.com/pytorch/pytorch/blob/d4990ad5a14fa2352ce3fff4f469cf499167c1a3/torch/_inductor/triton_heuristics.py#L903

peterbell10 commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -165,10 +166,56 @@ def precompile(self, warm_cache_only_with_cc=None):
+                    # 7.795ms to 4.883ms.
+                    if nreg <= 32:
+                        continue
+                    max_blocks_per_sm = 2048 // (32 * triton_config.num_warps)
+                    # TODO: this is hardcoded for A100, we should revise for H100
+                    # for optimal perf.
+                    if total_block <= max_blocks_per_sm * 108:
```
Could you use the device properties to make this more generic?
```suggestion
                    prop = torch.cuda.get_device_properties(self.meta["device"])
                    max_blocks_per_sm = prop.max_threads_per_multi_processor // (32 * triton_config.num_warps)
                    if total_block <= max_blocks_per_sm * prop.multi_processor_count:
```

I think we still need to hard code 32 registers per thread unfortunately.

eellison replied:
Can we compute 32 as `65536 / prop.max_threads_per_multi_processor` at least ? luckily, 65K registers per SM has stayed pretty stable

eellison reviewed (commented):
cool!

shunting314 commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -165,10 +166,56 @@ def precompile(self, warm_cache_only_with_cc=None):
+
+            if (
+                config.dynamic_scale_rblock
+                and self.heuristic_type == HeuristicType.REDUCTION
```
PERSISTENT_REDUCTION has an assumption that RBLOCK = next_power_of_two(rnumel).

If we want to reduce RBLOCK, we should convert PERSISTENT_REDUCTION to REDUCTION in the first place. The 'multi-kernel' work should help here since we can use autotune to pick between REDUCTION and PERSISTENT_REDUCTION.

shunting314 commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -165,10 +166,56 @@ def precompile(self, warm_cache_only_with_cc=None):
+                    if rblock <= 64:
+                        continue
+                    new_config = copy.deepcopy(triton_config)
+                    new_config.kwargs["RBLOCK"] = rblock // 2
```
yea, I'll do some dedup

shunting314 committed:
Update on "[inductor] scale down RBLOCK for occupancy"


For large reduction (with large xnumel and rnumel), we potentially need run large number of thread blocks. Occupancy matters here since with larger occupancy we can run more blocks on each SM and we may need less number of waves to run the entire kernel on the GPU.  Number of registers used by each thread can limit the occupancy. For A100, it's safe to say that register usage does not limit occupancy only if each thread use <= 32 registers. This PR leverage this observation and reduce RBLOCK (thus reduce registers used by each thread) if thread usage limit occupancy for large reduction.

The scenario mentioned can happen for the softmax kernel used in transformers. Here are some results get from devgpu:
- PLBartForCausalLM we improve from 1.88x (58.7ms) to 2.00x (55.82ms)
- TrOCRForCausalLM we improve from 1.45x (92.9ms) to 1.51x (89.12ms)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

shunting314 commented:
Here is a perf test [link](https://hud.pytorch.org/benchmark/compilers?startTime=Thu%2C%2007%20Sep%202023%2018%3A45%3A31%20GMT&stopTime=Thu%2C%2014%20Sep%202023%2018%3A45%3A31%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=shunting-scale-down-rblock&lCommit=f3a0be938587a318cb4e4651f8804bf50363da5a&rBranch=main&rCommit=e716505345f36c720232802e7e7a97321917d29a)
- HF 1.80x -> 1.82x .
- TB 1.67x -> 1.68x. It's not like noise since I see 
   - hf_GPT2_large 1.75x -> 1.88x
   - BERT_pytorch: 3.04x -> 3.11x
   - drq: 1.48x -> 1.53x
- TIMM 1.69x -> 1.70x . This can be noise.

Compilation time increases a bit
- HF 83s -> 87s
- TB 67s -> 68s
- TIMM 109s -> 111s

I hope we can trade the compilation time with some perf wins.

Chillee commented:
Why would there be compilation time increases?

eellison commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -165,10 +167,70 @@ def precompile(self, warm_cache_only_with_cc=None):
+                    if new_config in seen_configs:
+                        continue
+                    seen_configs.add(new_config)
+                    self.launchers.append(
```
We should remove a config from the launcher if we're replacing it with a better config so we don't need to benchmark it at runtime

shunting314 replied:
Check https://github.com/pytorch/pytorch/pull/109275#issuecomment-1720220714

shunting314 commented:
> Why would there be compilation time increases? Can we get runtime improvements without the compilation time increase?

That's because we generate extra configs to compile and benchmark. We could skip benchmarking an old config if we reduce its RBLOCK, but
- that may not change the overall compilation time much since I think compared to compilation, benchmarking is not that slow
- that may cause bad perf for some models. The changed config is not guaranteed to be better than the old config in every case. 

I discussed with Elias offline the idea to make the parallel compilation tasks more finer-grain.  Currently each tasks compiles all configs for a kernel. But maybe let each task compiles a single config can speedup things. Hopefully I can trade some compilation time here

shunting314 commented:
Hmm, I realized the extra configs are all get compiled by the main process (not any of the child process for parallel compiling). The reason is the parallel compiling subprocess don't have CUDA initialized. We can not load the cubin (in triton) so the number of registers are not available.

Not sure if there is anyway to workaround.

shunting314 commented:
> The reason is the parallel compiling subprocess don't have CUDA initialized. We can not load the cubin (in triton) so the number of registers are not available.

In theory there is one way to work around:

```
$ ptxas -arch sm_80 sumrow.ptx -v
ptxas info    : 0 bytes gmem
ptxas info    : Compiling entry function 'sumrow_kernel_0d1d23d' for 'sm_80'
ptxas info    : Function properties for sumrow_kernel_0d1d23d
    0 bytes stack frame, 0 bytes spill stores, 0 bytes spill loads
ptxas info    : Used 32 registers, 376 bytes cmem[0]
```

with '-v' ptxas will report the number of registers used.  But in inductor stack, ptxas is called by triton. We'll need to change triton to make num-used-registers information available without loading the cubin.

jansel reviewed (approved):
Maybe we should do this for other types of kernels too..

shunting314 commented:
@pytorchbot merge -f "test failure not related"
